snapshot/merge.go

```diff
@@ -0,0 +1,269 @@
+
+	if overlayBased && userns.RunningInUserNS() {
+		// When using an overlay-based snapshotter, if we are running rootless on a pre-5.11
+		// kernel, we will not have userxattr. This results in opaque xattrs not being visible
+		// to us and thus breaking the overlay-optimized differ. This also means that there are
+		// cases where in order to apply a deletion, we'd need to create a whiteout device but
+		// may not have access to one to hardlink, so we just fall back to not using hardlinks
+		// at all in this case.
+		userxattr, err := needsUserXAttr(ctx, sn, lm)
+		if err != nil {
+			bklog.G(ctx).Debugf("failed to check user xattr: %v", err)
+			useHardlinks = false
+		} else {
+			useHardlinks = userxattr
+		}
+	}
+
+	return &mergeSnapshotter{
+		Snapshotter:  sn,
+		lm:           lm,
+		useHardlinks: useHardlinks,
+		overlayBased: overlayBased,
+	}
+}
+
+func (sn *mergeSnapshotter) Merge(ctx context.Context, key string, parents []string, opts ...snapshots.Opt) error {
+	// Flatten the provided parent keys out. If any of them are them selves merges, don't use that key directly as
+	// a parent and instead use each of the merge inputs to that parent directly in the merge here. This is important
+	// for ensuring deletions defined between layers are preserved as part of each merge.
+	var mergeKeys []string
+	for _, parentKey := range parents {
+		parentInfo, err := sn.Stat(ctx, parentKey)
+		if err != nil {
+			return errors.Wrap(err, "failed to stat parent during merge")
+		}
+		if parentInfo.Kind != snapshots.KindCommitted {
+			return errors.Wrapf(err, "invalid kind %q for parent key %q provided to merge", parentInfo.Kind, parentKey)
+		}
+		if parentKeys := mergeKeysOf(parentInfo); parentKeys != nil {
+			mergeKeys = append(mergeKeys, parentKeys...)
+		} else {
+			mergeKeys = append(mergeKeys, parentKey)
+		}
+	}
+
+	// Map each key in mergeKeys to each key making up the snapshot chain.
+	// applyChains' outer slice has the same order as parents, which is expected to be lowest merge input to highest.
+	// applyChains' inner slice is ordered from highest layer to lowest.
+	var applyChains [][]string
+	var baseKey string
+	for i, parent := range mergeKeys {
+		if i == 0 && sn.overlayBased {
+			// Overlay-based snapshotters can skip the base of the merge and just use it as the parent of the merge snapshot.
+			// Other snapshotters will start empty (with baseKey set to "")
+			baseKey = parent
+			continue
+		}
+		var chain []string
+		for curkey := parent; curkey != ""; {
+			info, err := sn.Stat(ctx, curkey)
+			if err != nil {
+				return errors.Wrapf(err, "failed to stat chain key %q", curkey)
+			}
+			chain = append(chain, info.Name)
+			curkey = info.Parent
+		}
+		// add an empty key as the bottom layer so that the real bottom layer gets diffed with it and applied
+		chain = append(chain, "")
+		applyChains = append(applyChains, chain)
+	}
+
+	// Make the snapshot that will be merged into
+	prepareKey := identity.NewID()
```

Comment:
I don't understand how this optimizes out submerges that already have snapshots or even a duplicate merge. Especially because parent merges are flattened this seems quite important.

A possible solution might be to make these merge IDs deterministic with some sync primitives.

Reply:
Adding this layer might create conflicts with older data on upgrades. Take this into account if added later.